pennylane/devices/qubit/jaxpr_adjoint.py

```diff
@@ -0,0 +1,163 @@
+        # currently prevented by "no differentiable operator arithmetic."
+        # but better safe than sorry to keep this error
+        raise NotImplementedError  # pragma: no cover
+
+    if eqn.primitive.name != "expval_obs":
+        raise NotImplementedError("adjoint jvp only supports expectations of observables.")
+
+    mp = eqn.primitive.impl(*invals, **eqn.params)
+    bra = apply_operation(mp.obs, ket)
+    result = jnp.real(jnp.sum(jnp.conj(bra) * ket))
+    env[eqn.outvars[0]] = (result, None)
+    return bra
+
+
+def _other_prim_forward_pass(eqn: jax.core.JaxprEqn, env: dict) -> None:
+    """Handle any equation that is not an operator or measurement eqn.
+
+    Maps outputs back to the environment
+    """
+    invals, tangents = tuple(zip(*(_read(env, var) for var in eqn.invars)))
+    if eqn.primitive not in ad.primitive_jvps:
+        raise NotImplementedError(
+            f"Primitive {eqn.primitive} does not have a jvp rule and is not supported.."
+        )
+    outvals, doutvals = ad.primitive_jvps[eqn.primitive](invals, tangents, **eqn.params)
+    if not eqn.primitive.multiple_results:
+        outvals = [outvals]
+        doutvals = [doutvals]
+    for var, v, dv in zip(eqn.outvars, outvals, doutvals, strict=True):
+        env[var] = (v, dv)
+
+
+def _forward_pass(jaxpr: jax.core.Jaxpr, env: dict, num_wires: int):
+    """Calculate the forward pass of an adjoint jvp calculation."""
+    bras = []
+    ket = create_initial_state(range(num_wires))
+
+    for eqn in jaxpr.eqns:
+        if getattr(eqn.primitive, "prim_type", "") == "operator":
+            ket = _operator_forward_pass(eqn, env, ket)
+
+        elif getattr(eqn.primitive, "prim_type", "") == "measurement":
+            bra = _measurement_forward_pass(eqn, env, ket)
+            bras.append(bra)
+        else:
+            _other_prim_forward_pass(eqn, env)
+
+    results = [_read(env, var)[0] for var in jaxpr.outvars]
+    return bras, ket, results
+
+
+def _backward_pass(jaxpr, bras, ket, results, env):
+    """Calculate the jvps during the backward pass stage of an adjoint jvp."""
+    out_jvps = [jnp.zeros_like(r) for r in results]
+
+    modified = False
+    for eqn in reversed(jaxpr.eqns):
+        if getattr(eqn.primitive, "prim_type", "") == "operator" and isinstance(
+            eqn.outvars[0], jax.core.DropVar
+        ):
+            op = env[eqn.outvars[0]][0]
+
+            if eqn.invars:
+                t = _read(env, eqn.invars[0])[1]
+                if not isinstance(t, ad.Zero):
+                    disable()
+                    try:
+                        ket_temp = apply_operation(generator(op, format="observable"), ket)
+                    finally:
+                        enable()
+                    modified = True
+                    for i, bra in enumerate(bras):
+                        out_jvps[i] += -2 * t * jnp.imag(jnp.sum(jnp.conj(bra) * ket_temp))
```

Comment:
```suggestion
                tang = _read(env, eqn.invars[0])[1]
                if not isinstance(tang , ad.Zero):
                    disable()
                    try:
                        ket_temp = apply_operation(generator(op, format="observable"), ket)
                    finally:
                        enable()
                    modified = True
                    for i, bra in enumerate(bras):
                        out_jvps[i] += -2 * tang  * jnp.imag(jnp.sum(jnp.conj(bra) * ket_temp))
```

Can we add a more self-explanatory name than t? I prosed `tang`, but feel free to change it : )

Reply:
Just going with `tangent`. Thanks for the feedback :)